xmr-btc/tests/harness/storage.rs

```diff
@@ -0,0 +1,158 @@
+use anyhow::{anyhow, Context, Result};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::path::Path;
+
+pub struct Database<T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    db: sled::Db,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T> Database<T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    const LAST_STATE_KEY: &'static str = "latest_state";
+
+    pub fn open(path: &Path) -> Result<Self> {
+        let db =
+            sled::open(path).with_context(|| format!("Could not open the DB at {:?}", path))?;
+
+        Ok(Database {
+            db,
+            _marker: Default::default(),
+        })
+    }
+
+    pub async fn insert_latest_state(&self, state: &T) -> Result<()> {
+        let key = serialize(&Self::LAST_STATE_KEY)?;
+        let new_value = serialize(&state).context("Could not serialize new state value")?;
+
+        let old_value = self.db.get(&key)?;
+
+        self.db
+            .compare_and_swap(key, old_value, Some(new_value))
+            .context("Could not write in the DB")?
+            .context("Stored swap somehow changed, aborting saving")?;
+
+        self.db
+            .flush_async()
+            .await
+            .map(|_| ())
+            .context("Could not flush db")
+    }
+
+    pub fn get_latest_state(&self) -> anyhow::Result<T> {
+        let key = serialize(&Self::LAST_STATE_KEY)?;
+
+        let encoded = self
+            .db
+            .get(&key)?
+            .ok_or_else(|| anyhow!("State does not exist {:?}", key))?;
+
+        let state = deserialize(&encoded).context("Could not deserialize state")?;
+        Ok(state)
+    }
+}
+
+pub fn serialize<T>(t: &T) -> anyhow::Result<Vec<u8>>
+where
+    T: Serialize,
+{
+    Ok(serde_cbor::to_vec(t)?)
+}
+
+pub fn deserialize<T>(v: &[u8]) -> anyhow::Result<T>
+where
+    T: DeserializeOwned,
+{
+    Ok(serde_cbor::from_slice(&v)?)
+}
+
+#[cfg(test)]
+mod tests {
+    #![allow(non_snake_case)]
```

Comment:
I know it has not been introduced with this PR but I would welcome to rethink having var that are drastically different (public key vs private key) with names that only differ by casing (s vs S).

Devs use `priv` and `pub` prefixes.